File: src/main/java/caso3/lp2/tcpclient.py
import socket


class TcpClient:
    SERVER_PORT = 4444

    def __init__(self, ip_address, message_listener):
        self.server_ip = ip_address
        self.message_listener = message_listener
        self.out = None

    def send_message(self, message):
        try:
            if self.out is not None:
                self.out.write(message + '\n')
                self.out.flush()
        except Exception as e:
            print("TCP Client: Error sending message - ", e)

    def run(self):
        try:
            server_address = socket.gethostbyname(self.server_ip)
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
                sock.connect((server_address, self.SERVER_PORT))
                self.out = sock.makefile('w')
                in_stream = sock.makefile('r')
                while True:
                    server_message = in_stream.readline().strip()
                    if server_message and self.message_listener:
                        self.message_listener(server_message)
        except Exception as e:
            print("TCP Client: Error", e)
        

File: src/main/java/caso1/lp2/TCPClient.java
package caso1.lp2;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.InetAddress;
import java.net.Socket;

public class TCPClient {
    private final String serverIP;
    public static final int SERVER_PORT = 4455;
    private final OnMessageReceived messageListener;

    private PrintWriter out;

    public TCPClient(String ipAddress, OnMessageReceived listener) {
        serverIP = ipAddress;
        messageListener = listener;
    }

    public void sendMessage(String message) {
        if (out != null && !out.checkError()) {
            out.println(message);
            out.flush();
        }
    }

    public void run() {
        try {
            InetAddress serverAddress = InetAddress.getByName(serverIP);
            try (Socket socket = new Socket(serverAddress, SERVER_PORT)) {
                out = new PrintWriter(new BufferedWriter(new OutputStreamWriter(socket.getOutputStream())), true);
                BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
                while (true) {
                    String serverMessage = in.readLine();
                    if (serverMessage != null && messageListener != null) {
                        messageListener.messageReceived(serverMessage);
                    }
                }
            } catch (Exception e) {
                System.out.println("TCP Client: Error " + e);
            }
        } catch (Exception e) {
            System.out.println("TCP Client: Error " + e);
        }
    }

    public interface OnMessageReceived {
        void messageReceived(String message);
    }
}
